WIP: Database and table visualization

Erik-Handeland opened the pull request:
# Description

Implements [Datastore](https://cloud.google.com/appengine/docs/standard/java/datastore/api-overview) into the project along with allowing the upload of mock data via CSV file. All datastore entries are uploaded through the SetDatastore function and retrieved via the GetDatastore function. The data is visualized to the end user by accessing the /tables page on the project website.

# Checklist:

- [x] Code follows the style guidelines
- [x] Code commented in relevant places
- [x] Changes to necessary documentation  
- [x] Basic manual testing performed after deploying app

Commit:
Added hardcode Json table

Erik-Handeland committed:
Switch from local JSON to dataStore

Erik-Handeland committed:
added flagged option to datastore form

Erik-Handeland committed:
Quick add button for random table data

Erik-Handeland committed:
replaced mock data array with read from CSV file

Erik-Handeland committed:
chaged datastore to hold freq. attribute

Erik-Handeland committed:
Abstracted database upload and retrieval into seperate classes

arunk054 commented:
Too many fines, can you may be synch up with Melba's changes?

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+//PCAP storage class
+public class GetDatastore {
+
+    public ArrayList<PCAPdata> dataTable = new ArrayList<PCAPdata>();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+  public GetDatastore() 
```
Indentation is bad. Are you using some sort of Editor?

Erik-Handeland replied:
Was doing most of my coding on Google cloud shell, not the best editor, Switch to personal laptop with VS code and reformatted code

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+public class GetDatastore {
+
+    public ArrayList<PCAPdata> dataTable = new ArrayList<PCAPdata>();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```
These two should go in the constructor even though it would not matter much but for readability.
And they should be private.

Erik-Handeland replied:
removed dataTable, ran into issues moving datastore into constructor, switch it to private

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+//PCAP storage class
+public class GetDatastore {
```
Usually a class name is a Noun. So something like PCAPDataStore? I am not sure what this class is trying to achieve so add some comments. If someone instantiates this they can get a list of PCAPdata objects per file or for all files?

arunk054 replied:
Make this an interface, something like `GenericPCAPDao` which has methods like getPCAPObjects(id).. you can then implement the interface with your datastore specific implementation.

Erik-Handeland replied:
Changed class/method names

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+
+
+  //get all properties in the entity "data" 
+  public void getAll(){
```
a method something like getALL should return something. So just return the arraylist instead of populating it.

arunk054 replied:
Probably you wont need this method since the client can just call getEntity(PCAPInterface.DATASTORE_NAME);

Erik-Handeland replied:
Method was removed

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+      PCAPdata temp = new PCAPdata(source, destination, domain, 
+      location,protocol, size, flagged, frequency);
+       
+      dataTable.add(temp);
```
create a temporary variable and return this list.

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+
+  //get all properties in the entity "data" 
+  public void getAll(){
+    Query query = new Query("data").addSort("Flagged", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
+
+    for (Entity entity : results.asIterable()) {
+
+      String source = (String) entity.getProperty("Source");
+      String destination = (String) entity.getProperty("Destination");
+      String domain = (String) entity.getProperty("Domain");
+      String location = (String) entity.getProperty("Location");
+      String protocol = (String) entity.getProperty("Protocol");
+      int size = (int) (long) entity.getProperty("Size");
+      boolean flagged = (Boolean) entity.getProperty("Flagged");
+      int frequency = (int) (long) entity.getProperty("Frequency");
+
+      PCAPdata temp = new PCAPdata(source, destination, domain, 
+      location,protocol, size, flagged, frequency);
```
all of this can be reduced to one line if you just call your getEntity("data") right?

Erik-Handeland replied:
Method was removed

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+//PCAP storage class
+public class GetDatastore {
+
+    public ArrayList<PCAPdata> dataTable = new ArrayList<PCAPdata>();
```
you dont need this variable. you should not maintain this state unless you are using it to cache and avoid looking up the datastore every time.

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+
+  //get all properties in the entity "data" 
+  public void getAll(){
+    Query query = new Query("data").addSort("Flagged", SortDirection.DESCENDING);
```
puth the string "data"" in a Constant variable.

Erik-Handeland committed:
removed vendor folder

arunk054 commented on src/main/java/com/google/sps/datastore/GetDatastore.java:
```diff
@@ -0,0 +1,74 @@
+
+
+  //get all properties in a given entity (have each pcap files as seperate entity in datastore)
+  public void getEntity(String searchEntity){
```
entity is very generic, basically you are trying to get pcap objects per file, List<> getPCAPObjects(String file_id)

Erik-Handeland commented:
Removed the vendor folder which was causing the extra files.

arunk054 commented on src/main/java/com/google/sps/datastore/SetDatastore.java:
```diff
@@ -0,0 +1,49 @@
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+//PCAP storage class
+public class SetDatastore {
```
Actually combine your Get and Set in the same interface. I was not clear when you mentioned yesterday but the upload and download are basically in the same DAO. Your GenericPCAPDao interface wil have these two methoids and you can implement them in a class.

Erik-Handeland committed:
Combind get/set Datastore

Combind seperate datastore classes and reformated code to match google style guide.

arunk054 commented:
I tried to change the base to Melba's branch and it shows there is no diff, so looks like all of this is already merged to Melba. So we just close it then.

Erik-Handeland committed:
Const for filename

Erik-Handeland committed:
Added datastore interface

Erik-Handeland committed:
Seperated mockData servlet into seperate MockDataLoader class

Erik-Handeland committed:
Update data.csv

Erik-Handeland committed:
modified CSV loader

Erik-Handeland commented:
Datastore Attributes

    | Key | ID | Sources | Destination | Protocol | Size |
    |-----|----|---------|-------------|----------|------|
    |     |    | String  | String      | String   | Long |
